### PR TITLE
Added option to templatetags to use the full url

### PR DESCRIPTION
**Pull request by trumpet2012**
This PR is in reference to #143.

I added the ability to specify at the templatetag level whether or not to use the full url when creating menu items.

This adds the ability to do: `{% main_menu use_absolute_urls=True %}`

## Upgrade considerations
- ~~Signature of `get_repeated_menu_item` and `modify_submenu_items` for MenuPageMixin now includes the kwarg `use_absolute_urls`, which defaults to False. Anybody who overrode either of these methods will need to update their method signature.~~ This is addressed by checking the method signatures and raising a deprecation warning if the kwarg is not present.

## TODO:
- [x] Finish tests
- [x] Update documentation

### CODE REVIEW

File: wagtailmenus/models/menuitems.py [hunk 1]
```diff
@@ -96,6 +96,16 @@ def relative_url(self, site=None):
                 return ''
         return self.link_url + self.url_append
 
+    def get_full_url(self):
+        if self.link_page:
+            try:
+                return self.link_page.full_url + self.url_append
```
Comment (ababic) [inline wagtailmenus/models/menuitems.py]: I feel as though this method would benefit from accepting a `request` parameter, and passing that to the new `link_page.get_full_url()` method if the page has it.

File: wagtailmenus/models/menuitems.py [hunk 2]
```diff
@@ -96,6 +96,16 @@ def relative_url(self, site=None):
...
+            except TypeError:
+                return ''
+        return self.link_url + self.url_append
+
+    full_url = property(get_full_url)
```
Comment (ababic) [inline wagtailmenus/models/menuitems.py]: If we accept a request parameter as outlined above, declaring a property method probably isn't necessary

File: wagtailmenus/models/pages.py [hunk 3]
```diff
@@ -100,14 +101,18 @@ def has_submenu_items(self, current_page, allow_repeating_parents,
...
+        original_menu_tag, request=None, use_absolute_urls=False,
     ):
         """Return something that can be used to display a 'repeated' menu item
         for this specific page."""
 
         menuitem = copy(self)
         setattr(menuitem, 'text', self.repeated_item_text or self.title)
-        setattr(menuitem, 'href', self.relative_url(current_site))
+        if use_absolute_urls:
+            url = self.full_url
```
Comment (ababic) [inline wagtailmenus/models/pages.py]: This would become:

`url = self.get_full_url(request)`

**Comment by ababic** [comment]
Hi @trumpet2012. Thanks for taking the time to put this PR together. I requested some changes above. I'd also like to avoid introducing any backwards-incompatible changes here if possible. We could probably do something using `wagtail.wagtailcore.utils.accepts_kwarg` and raise a `DeprecationWarning`

**Comment by ababic** [comment]
Hey @trumpet2012. I could've been clearer in my requests :) It's okay for the `main_menu`, `flat_menu`, `children_menu` and `section_menu` tags to define `True/False` default values for `use_absolute_urls` - It's just `sub_menu` that should be altered to pick up the value up from the context :)

**Comment by trumpet2012** [comment]
@ababic That makes sense, I'll put those back.

File: wagtailmenus/templatetags/menu_tags.py [hunk 4]
```diff
@@ -228,7 +235,8 @@ def get_sub_menu_items_for_page(
 @register.simple_tag(takes_context=True)
 def sub_menu(
     context, menuitem_or_page, stop_at_this_level=None, use_specific=None,
-    allow_repeating_parents=None, apply_active_classes=None, template=''
+    allow_repeating_parents=None, apply_active_classes=None, template='',
+    use_absolute_urls=False,
```
Comment (trumpet2012) [inline wagtailmenus/templatetags/menu_tags.py]: You're right, this is a much better way. I've updated it accordingly.

File: wagtailmenus/templatetags/menu_tags.py [hunk 5]
```diff
@@ -347,15 +358,20 @@ def section_menu(
...
     )
 
     """
     We want `section_root` to have the same attributes as primed menu
     items, so it can be used in the same way in a template if required.
     """
     setattr(section_root, 'text', section_root.title)
-    setattr(section_root, 'href', section_root.relative_url(site))
+    if use_absolute_urls:
+        url = section_root.full_url
```
Comment (trumpet2012) [inline wagtailmenus/templatetags/menu_tags.py]: Yeah it looks like it would be better for performance, since they use the request to cache the root path lookup.

File: wagtailmenus/models/menuitems.py (hunk 1 above)
Comment (trumpet2012) [inline wagtailmenus/models/menuitems.py]: I updated it, make sure it looks good to you.

File: wagtailmenus/tests/test_models.py [hunk 6]
```diff
@@ -276,3 +279,22 @@ def test_linkpage_redirects_when_served(self):
             response,
             '/superheroes/marvel-comics/spiderman/?somevar=value'
         )
+
+
+class TestNoAbsoluteUrlsPage(TestCase):
+    fixtures = ['updated_test.json']
```
Comment (trumpet2012) [inline wagtailmenus/tests/test_models.py]: I added a new model to trigger the `DeprecationWarning`s. Wasn't sure how you want to handle updating fixtures, so I generated a new fixture. Tried to limit it to the apps/models that the are in `test.json`. Generated it using 
```
python manage.py dumpdata tests wagtailmenus sites wagtailcore.site wagtailcore.page -a --indent 4 --natural-foreign --natural-primary
```

**Comment by trumpet2012** [comment]
@ababic Not sure why the coverage has dropped so dramatically, have any ideas? Ran the coverage report locally and it reported 98% coverage.

**Comment by ababic** [comment]
Hi @trumpet2012. The coverage report looks buggy to me. I guessing it will resolve itself next time changes are committed. This is looking really good at a first glance, but I'll need some time to review it all properly, which I hope to do soon. Thanks!

File: wagtailmenus/tests/templates/menus/custom_sub_menu.html [hunk 7]
```diff
@@ -5,7 +5,7 @@
 	    <li class="{{ item.active_class }}{% if item.extra_menuitem_css_class %} {{ item.extra_menuitem_css_class }}{% endif %}">
 	        <a href="{{ item.href }}">{{ item.text }}</a>
 	        {% if item.has_children_in_menu %}
-	            {% sub_menu item %}
+	            {% sub_menu item%}
```
Comment (ababic) [inline wagtailmenus/tests/templates/menus/custom_sub_menu.html]: Could we revert this change?

File: wagtailmenus/tests/test_models.py (hunk 6 above)
Comment (ababic) [inline wagtailmenus/tests/test_models.py]: Thanks, this is very thorough! However, recently I've been leaning towards having specific test cases like this create their own one-off pages to test with. Fixtures are well suited for some things (like efficiently populating a page tree with a reliable structure to render menus with), but less useful / clear for more 'unit-test like' things. Here's how I did the tests for the new AbstractLinkPage, if you'd like to see an example: https://github.com/rkhleics/wagtailmenus/blob/master/wagtailmenus/tests/test_models.py#L146

File: wagtailmenus/tests/templates/menus/custom_sub_menu.html (hunk 7 above)
Comment (trumpet2012) [inline wagtailmenus/tests/templates/menus/custom_sub_menu.html]: Fixed the space

File: wagtailmenus/templatetags/menu_tags.py [hunk 8]
```diff
@@ -608,8 +645,17 @@ def prime_menu_items(
             menuitem.link_page = page
 
         # Both `Page` and `MenuItem` objects have a `relative_url` method
-        # we can use to calculate a value for the `href` attribute.
-        setattr(item, 'href', item.relative_url(current_site))
+        # and a `get_full_url` method (added in Wagtail 1.11) or `full_url`
```
Comment (ababic) [inline wagtailmenus/templatetags/menu_tags.py]: Talking about `Page` and `MenuItem` objects in the same comment here is a little confusing now, as the wagtail version stuff applies to one and not the other. I think we should probably remove this all together, and replace it with two shorter comments.

And one above what is currently at line `652`:

> Pages only have `get_full_url` from Wagtail 1.11 onwards

And one above what is currently at line `655`:

> Fallback for Wagtail versions prior to 1.11

File: wagtailmenus/tests/test_models.py (hunk 6 above)
Comment (trumpet2012) [inline wagtailmenus/tests/test_models.py]: Yeah that makes sense. I have taken it out.

File: wagtailmenus/templatetags/menu_tags.py (hunk 8 above)
Comment (trumpet2012) [inline wagtailmenus/templatetags/menu_tags.py]: Yeah that is a lot clearer.

File: README.md [hunk 9]
```diff
@@ -412,6 +419,7 @@ The `{% sub_menu %}` tag is used within menu templates to render additional leve
...
 - **`template`** (default: `''`): Lets you render the menu to a template of your choosing. If not specified, the `sub_menu_template` value added to the context by the original menu tag will be used.
 - **`use_specific`**: By default, the tag will inherit this behaviour from whatever was specified for the original menu tag. However, the value can be overridden by passing this option to the {% sub_menu %} tag in your custom menu template. Take a look at the [Specific pages instances and performance](#specific-page-use) section below for a description of the option values supported.
+- **`use_absolute_urls`** (default: `None`): Will use absolute URLs when creating the links for pages. When `None` is specified we check the template context for `use_absolute_urls`, this allows the sub menu to honor the value specified for any parent menus. If `use_absolute_urls` is not in the context and not passed to the tag, it defaults to `False`.
```
Comment (ababic) [inline README.md]: And could this one be something like:

> By default, the tag will inherit this behaviour from whatever was specified for the original menu tag. However, this can be overridden by explicitly adding `use_absolute_urls=True` or `use_absolute_urls=False` to a `{% sub_menu %}` tag in a custom menu template. If `True`, absolute page URLs will be used in the menu instead a relative URLs.

File: README.md [hunk 10]
```diff
@@ -424,6 +432,7 @@ The following variables are added to the context by all of the above tags, which
 - **`max_levels`**: The maximum number of levels that should be rendered, as determined by the original `main_menu`, `section_menu`, `flat_menu` or `children_menu` tag call.
 - **`allow_repeating_parents`**: A boolean indicating whether `MenuPage` fields should be respected when rendering further menu levels.
 - **`apply_active_classes`**: A boolean indicating whether `sub_menu` tags should attempt to add  'active' and 'ancestor' classes to menu items when rendering further menu levels.
+- **`use_absolute_urls`**: A boolean indicating whether menu items should be linked using absolute URLs.
```
Comment (ababic) [inline README.md]: Could this be reworded to:

> A boolean indicating whether absolute page URLs should be used for menu item links when rendering.

**Comment by ababic** [comment]
@trumpet2012 Also, don't worry about adding anything to the CHANGELOG - I tend to do all that separately before deployment. If you just let me know here who I need to credit there (Just yourself, or your company, or a client name if they sponsored the feature), I will put something appropriate together.

File: README.md [hunk 11]
```diff
@@ -163,6 +163,7 @@ The `{% main_menu %}` tag allows you to display the `MainMenu` defined for the c
 - **`apply_active_classes`** (default: `True`): The tag will attempt to add 'active' and 'ancestor' CSS classes to the menu items (where applicable) to indicate the active page and ancestors of that page. To disable this behaviour, add `apply_active_classes=False` to the tag in your template. You can change the CSS classes used by adding `WAGTAILMENUS_ACTIVE_CLASS` and `WAGTAILMENUS_ACTIVE_ANCESTOR_CLASS` settings to your project's settings module.
 - **`template`** (default: `''`): Lets you render the menu to a template of your choosing. If not provided, wagtailmenus will look in several locations for an appropriate template (see below for details).
 - **`sub_menu_template`** (default: `''`): Lets you specify a template to be used for rendering sub menus. All subsequent calls to `{% sub_menu %}` within the context of the section menu will use this template unless overridden by providing a `template` value to `{% sub_menu %}` in a menu template. If not provided, wagtailmenus will look in several locations for an appropriate template (see below for details).
+- **`use_absolute_urls`** (default: `False`): Will use absolute URLs when creating the links for pages.
```
Comment (trumpet2012) [inline README.md]: It's been updated.

**Comment by trumpet2012** [comment]
@ababic Thanks for looking at this. I know it takes a lot of your time to go through all of this. As for the changelog, if you could credit me and my company (Morris Technology) that would be awesome.

**Comment by trumpet2012** [comment]
Coverage is still showing a huge drop. Running locally still shows 98% coverage:
```
Name                                                       Stmts   Miss  Cover
------------------------------------------------------------------------------
wagtailmenus/__init__.py                                      18      4    78%
wagtailmenus/app_settings.py                                  97      0   100%
wagtailmenus/apps.py                                           4      0   100%
wagtailmenus/context_processors.py                            35      0   100%
wagtailmenus/development/__init__.py                           0      0   100%
wagtailmenus/development/urls.py                               8      8     0%
wagtailmenus/forms.py                                         14      0   100%
wagtailmenus/managers.py                                       5      0   100%
wagtailmenus/migrations/0001_initial.py                        6      0   100%
wagtailmenus/migrations/0002_auto_20160129_0901.py             5      0   100%
wagtailmenus/migrations/0003_auto_20160129_2339.py             5      0   100%
wagtailmenus/migrations/0004_auto_20160130_0024.py             5      0   100%
wagtailmenus/migrations/0005_auto_20160130_2209.py             5      0   100%
wagtailmenus/migrations/0006_auto_20160131_1347.py             5      0   100%
wagtailmenus/migrations/0007_auto_20160131_2000.py             5      0   100%
wagtailmenus/migrations/0008_auto_20160131_2327.py             5      0   100%
wagtailmenus/migrations/0009_auto_20160201_0859.py             5      0   100%
wagtailmenus/migrations/0010_auto_20160201_1558.py             5      0   100%
wagtailmenus/migrations/0011_auto_20160415_1519.py             5      0   100%
wagtailmenus/migrations/0012_auto_20160419_0039.py             6      0   100%
wagtailmenus/migrations/0013_auto_20160423_1124.py             5      0   100%
wagtailmenus/migrations/0014_auto_20160423_1339.py             5      0   100%
wagtailmenus/migrations/0015_auto_20160423_1348.py             5      0   100%
wagtailmenus/migrations/0016_auto_20160930_0645.py             5      0   100%
wagtailmenus/migrations/0017_auto_20161013_1658.py             6      0   100%
wagtailmenus/migrations/0018_auto_20161204_2043.py             5      0   100%
wagtailmenus/migrations/0019_auto_20161204_2143.py             5      0   100%
wagtailmenus/migrations/0020_auto_20161210_0004.py             6      0   100%
wagtailmenus/migrations/0021_auto_20170106_2352.py             5      0   100%
wagtailmenus/migrations/__init__.py                            0      0   100%
wagtailmenus/models/__init__.py                                3      0   100%
wagtailmenus/models/menuitems.py                              71      5    93%
wagtailmenus/models/menus.py                                 182      2    99%
wagtailmenus/models/pages.py                                 132     10    92%
wagtailmenus/panels.py                                         9      0   100%
wagtailmenus/settings/__init__.py                              0      0   100%
wagtailmenus/settings/base.py                                 23      0   100%
wagtailmenus/settings/development.py                          12     12     0%
wagtailmenus/settings/testing.py                               9      0   100%
wagtailmenus/templatetags/__init__.py                          0      0   100%
wagtailmenus/templatetags/menu_tags.py                       213      2    99%
wagtailmenus/tests/__init__.py                                 0      0   100%
wagtailmenus/tests/migrations/0001_initial.py                  7      0   100%
wagtailmenus/tests/migrations/0002_homepage.py                 6      0   100%
wagtailmenus/tests/migrations/0003_auto_20160418_1204.py       5      0   100%
wagtailmenus/tests/migrations/0004_auto_20160419_2022.py       5      0   100%
wagtailmenus/tests/migrations/0005_auto_20160419_2025.py       5      0   100%
wagtailmenus/tests/migrations/0006_auto_20160423_1339.py       5      0   100%
wagtailmenus/tests/migrations/0007_auto_20160423_1348.py       5      0   100%
wagtailmenus/tests/migrations/0008_contactpage.py              6      0   100%
wagtailmenus/tests/migrations/0009_typicalpage.py              6      0   100%
wagtailmenus/tests/migrations/0010_auto_20161211_2116.py       9      0   100%
wagtailmenus/tests/migrations/0011_auto_20170106_2355.py       5      0   100%
wagtailmenus/tests/migrations/0012_linkpage.py                 6      0   100%
wagtailmenus/tests/migrations/0013_auto_20170630_1002.py       5      0   100%
wagtailmenus/tests/migrations/0014_noabsoluteurlspage.py       6      0   100%
wagtailmenus/tests/migrations/__init__.py                      0      0   100%
wagtailmenus/tests/models/__init__.py                          2      0   100%
wagtailmenus/tests/models/menus.py                            34      0   100%
wagtailmenus/tests/models/pages.py                            58      1    98%
wagtailmenus/tests/models/utils.py                            13      0   100%
wagtailmenus/tests/test_backend.py                           145      0   100%
wagtailmenus/tests/test_custom_models.py                     143      0   100%
wagtailmenus/tests/test_menu_rendering.py                    279      0   100%
wagtailmenus/tests/test_models.py                            143      0   100%
wagtailmenus/tests/urls.py                                     5      0   100%
wagtailmenus/utils/__init__.py                                 0      0   100%
wagtailmenus/utils/deprecation.py                              8      0   100%
wagtailmenus/utils/inspection.py                              12      2    83%
wagtailmenus/utils/misc.py                                    29      3    90%
wagtailmenus/utils/template.py                                27      0   100%
wagtailmenus/utils/version.py                                 13      2    85%
wagtailmenus/views.py                                         95      0   100%
wagtailmenus/wagtail_hooks.py                                 78      0   100%
------------------------------------------------------------------------------
TOTAL                                                       2109     51    98%
```

File: README.md (hunk 11 above)
Comment (ababic) [inline README.md]: Sorry, I don't think I was clear enough in my explaination there. it looks as though you've left the reference to `{% main_menu %}` the same for each tag. That needs to change for each tag the instructions are for.

**Comment by ababic** [comment]
Hi @trumpet2012, just a few more documentation changes and we should be there. Once completed, is there any chance you can submit as a new PR with the changes squashed into a single commit, or a few sensibly focussed ones? That should hopefully fix the coverage report and allow me to get it merged into master.

File: README.md (hunk 11 above)
Comment (trumpet2012) [inline README.md]: No that's my fault, I should have caught that. I've updated them.

**Comment by trumpet2012** [comment]
@ababic Yeah thats no problem. Once we get everything fixed, I'll squash the commits and open a new PR.

**Comment by ababic** [comment]
@trumpet2012 Okay, great. If you can do the squashing now and submit another PR, i'll give things another look over there, and hopefully get this merged soon.

**Comment by ababic** [comment]
superseded by #147
